### ISSUE TITLE
Errors in processEnded can cause processes to be eternally reaped

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3116]
|<img alt="jml-insecure's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jml-insecure reported|
|-|-|
|Trac ID|trac#3116|
|Type|defect|
|Created|2008-03-25 00:36:34Z|
|Branch|https://github.com/twisted/twisted/tree/process-ended-3116-2|

`_BaseProcess.reapProcess` has some code that looks like this:

```python
        if pid:
            self.processEnded(status)
            unregisterReapProcessHandler(pid, self)
```

`processEnded` calls `processEnded` on the `ProcessProtocol` without any protection. If the protocol raises an exception, `unregisterReapProcessHandler` will never be called.

<details><summary>Searchable metadata</summary>

```
trac-id__3116 3116
type__defect defect
reporter__jml_insecure jml-insecure
priority__low low
milestone__ 
branch__branches_process_ended_3116_2 branches/process-ended-3116-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1206405394000000 1206405394000000
changetime__1214489049000000 1214489049000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

(In [23076]) Branching to 'process-ended-3116'

**Comment by twisted-trac** [comment]
|<img alt="jml-insecure's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|jml-insecure removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="jml-insecure's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|jml-insecure commented|
|-|-|

This bug is actually invalid -- Twisted behaves correctly here.

I've added a test that demonstrates it. It was tricky to write, so maybe it's worth adding to Twisted.

This branch also moves the process tests to internet/test.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

I haven't look thoroughly, but the buildbot shows that the new test is failing under windows.

**Comment by twisted-trac** [comment]
|<img alt="jml-insecure's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|jml-insecure commented|
|-|-|

That's not entirely surprising. What's the failure?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

http://buildbot.twistedmatrix.com/builders/winxp32-py2.4-select/builds/251/steps/select/logs/problems

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @jml|
|-|-|

 * the return value of flushLoggedErrors should be checked
 * the docstring of test_errorInProcessEnded should not use 'We'
 * the test must be fixed or skipped under windows.

The move of everything makes it hard to review :/. I also wonder if the script should not be removed, and a temporary file created instead. This way, all the content of the tests would in test_process, not split between test_process and the script. That's out of scope of this one though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [23995]) Branching to 'process-ended-3116-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun removed owner|
|-|-|

I moved everything back, so it should be easier to review this time.  I also got rid of `process_trivial.py` and re-used `process_twisted.py` which serves just as well.  I checked the return value of `flushLoggedErrors`, adjusted the test method docstring, and moved the test to a posix-specific testcase, so it won't run on Windows (where it makes no sense - it is testing posix-specific implementation details).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set owner to @exarkun|
|-|-|

THe test looks great. I'm a bit worried about adding win32eventreactor to the reactormixins: I think it doesn't work currently under Windows, and might even froze the reactor. FWIW, I'm +1 for not putting it at all in the tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @therve|
|-|-|

I can comment it out.  It'd be great to add, but I don't really care about it right now, I just fixed the name because I noticed it was wrong.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set owner to @exarkun|
|-|-|

Apparently the builders were able to run that, so please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [24056]) Merge process-ended-3116-2

Author: jml, exarkun
Reviewer: therve
Fixes: [#3116](https://github.com/twisted/twisted/issues/3116)

Add a unit test for the handling of exceptions from
`IProcessProtocol.processEnded` implementations.  Make
sure the error is logged and reactor state associated
with it is still cleaned up.

Also fix the name of the Windows Event reactor in the
reactor builder harness so that it is possible for that
reactor to be tested on platforms where it works.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
